query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs

```diff
@@ -231,6 +236,66 @@ impl ChainingHashTable {
         Ok(results)
     }
 
+    // Todo(xudong963): optimize the performance of this function
+    fn filter_block(&self, probe_block: &DataBlock, merged_block: DataBlock) -> Result<DataBlock> {
+        let mut result_block = merged_block;
+        for scalar in self.other_conditions.iter() {
+            let func_ctx = self.ctx.try_get_function_context()?;
+            let eval = ScalarEvaluator::try_create(scalar)?;
+            // `predicate_vector` contains a column, which is a boolean column.
+            let predicate_vector = eval.eval(&func_ctx, &result_block)?;
+            // Here, we directly use `predicate_vector` to filter the result block.
+            // But pay attention to the fact that **reserved side** may also be filtered.
+            // **reserved side** is the left side in left join, and the right side in right join.
+            result_block = DataBlock::filter_block(&result_block, predicate_vector.vector())?;
+        }
```

Comment:
Maybe you can do this in `probe_join`?

Reply:
Your point is that don't need to write this logic again in the `probe` function?

Reply:
https://github.com/datafuselabs/databend/blob/5e73d12f8b03b53f22a5f99fe70361515a57466a/query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs#L341-L352

I think it's fine to filter the `block` gathered from `RowSpace` with non-equi predicates, and then check if there is necessary to append `NULL`s for it.

So you don't need to do the costly `except_block` operation.

Reply:
> For condition such as `t1.a > t2.c`, it need the `merged_block` to be filtered

Yes, but the `merge_block` is invoked right after this.

https://github.com/datafuselabs/databend/blob/5e73d12f8b03b53f22a5f99fe70361515a57466a/query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs#L366